File: src/ai/memory/session.service.ts
// ============================================================
// PATH: backend/src/ai/memory/session.service.ts
// PURPOSE: Stores and retrieves agent session from Redis
//          Each session holds stage, history, project data
// ============================================================

import redis from '../../config/redis'
import { AgentSession } from '../shared/agent.types'
import { AgentStage, SESSION_PREFIX, SESSION_TTL_SECONDS } from '../shared/constants'
import { v4 as uuidv4 } from 'uuid'

export class SessionService {

  // ── Build Redis key from sessionId ───────────────────────
  private key(sessionId: string): string {
    return `${SESSION_PREFIX}${sessionId}`
  }

  // ── Create a brand new session ───────────────────────────
  async create(clientName?: string, clientId?: string): Promise<AgentSession> {
    const session: AgentSession = {
      sessionId: uuidv4(),
      clientId,
      clientName: clientName || 'Client',
      stage: AgentStage.UNDERSTAND,
      history: [],
      project: {},
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    }

    await this.save(session)
    return session
  }

  // ── Get existing session by ID ───────────────────────────
  async get(sessionId: string): Promise<AgentSession | null> {
    try {
      const raw = await redis.get(this.key(sessionId))
      if (!raw) return null
      return JSON.parse(raw) as AgentSession
    } catch (err) {
      console.error('❌ SessionService.get error:', err)
      return null
    }
  }

  // ── Save / update session ────────────────────────────────
  async save(session: AgentSession): Promise<void> {
    try {
      session.updatedAt = new Date().toISOString()
      await redis.set(
        this.key(session.sessionId),
        JSON.stringify(session),
        { EX: SESSION_TTL_SECONDS }
      )
    } catch (err) {
      console.error('❌ SessionService.save error:', err)
      throw err
    }
  }

  // ── Update stage ─────────────────────────────────────────
  async updateStage(sessionId: string, stage: AgentStage): Promise<void> {
    const session = await this.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)
    session.stage = stage
    await this.save(session)
  }

  // ── Append message to history ────────────────────────────
  async appendMessage(
    sessionId: string,
    role: 'user' | 'assistant',
    content: string
  ): Promise<void> {
    const session = await this.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)
    session.history.push({ role, content })
    await this.save(session)
  }

  // ── Update extracted project data ────────────────────────
  async updateProject(sessionId: string, project: object): Promise<void> {
    const session = await this.get(sessionId)
    if (!session) throw new Error(`Session not found: ${sessionId}`)
    session.project = { ...session.project, ...project }
    await this.save(session)
  }

  // ── Delete session (cleanup) ─────────────────────────────
  async delete(sessionId: string): Promise<void> {
    await redis.del(this.key(sessionId))
  }

  // ── Get or create session ────────────────────────────────
  async getOrCreate(sessionId?: string, clientName?: string, clientId?: string): Promise<AgentSession> {
    if (sessionId) {
      const existing = await this.get(sessionId)
      if (existing) return existing
    }
    return this.create(clientName, clientId)
  }
}

File: src/ai/memory/session_service.py
import json
import uuid
from datetime import datetime, timedelta
from typing import Any, Dict, Optional, List

from shared.constants import AgentStage
from shared.db import Database
from shared.redis_client import RedisClient # New import

class SessionService:
    def __init__(self, db=None):
        self._db = db
        self.REDIS_TTL = 3600 # 1 hour for active sessions

    async def get_db(self):
        if self._db is None:
            self._db = await Database.get_db()
        return self._db

    async def create(self, client_name: str = "Client", client_id: str = None) -> Dict[str, Any]:
        session_id = str(uuid.uuid4())
        session = {
            "sessionId": session_id,
            "clientId": client_id,
            "clientName": client_name,
            "stage": AgentStage.UNDERSTAND,
            "history": [],
            "project": {},
            "createdAt": datetime.now().isoformat(),
            "updatedAt": datetime.now().isoformat(),
        }
        await self.save(session)
        return session

    async def get(self, session_id: str) -> Optional[Dict[str, Any]]:
        # 1. Try REDIS (Short-term)
        try:
            redis = await RedisClient.get_client()
            cached = await redis.get(f"ai_session:{session_id}")
            if cached:
                print(f"⚡ Session from Redis: {session_id}")
                return json.loads(cached)
        except Exception as e:
            print(f"⚠️ Redis get failed: {e}")

        # 2. Try MONGODB (Long-term)
        db = await self.get_db()
        session = await db.ai_sessions.find_one({
            "$or": [
                {"sessionId": session_id},
                {"_id": session_id}
            ]
        })
        
        if session and "_id" in session:
            session["_id"] = str(session["_id"])
            # Cache back to Redis
            try:
                redis = await RedisClient.get_client()
                await redis.setex(f"ai_session:{session_id}", self.REDIS_TTL, json.dumps(session, default=str))
            except: pass

        return session

    async def save(self, session: Dict[str, Any]):
        db = await self.get_db()
        session_id = session.get("sessionId")
        session["updatedAt"] = datetime.now().isoformat()
        
        # 1. Save to MONGODB
        to_mongo = {**session}
        if "_id" in to_mongo: del to_mongo["_id"]
        await db.ai_sessions.update_one(
            {"sessionId": session_id},
            {"$set": to_mongo},
            upsert=True
        )

        # 2. Save to REDIS
        try:
            redis = await RedisClient.get_client()
            await redis.setex(f"ai_session:{session_id}", self.REDIS_TTL, json.dumps(session, default=str))
        except Exception as e:
            print(f"⚠️ Redis save failed: {e}")

    async def get_or_create(self, session_id: str = None, client_name: str = "Client", client_id: str = None) -> Dict[str, Any]:
        if session_id:
            existing = await self.get(session_id)
            if existing:
                return existing
        return await self.create(client_name, client_id)

    async def update_stage(self, session_id: str, stage: AgentStage):
        db = await self.get_db()
        await db.ai_sessions.update_one(
            {"sessionId": session_id},
            {"$set": {"stage": stage, "updatedAt": datetime.now()}}
        )

    # 🧠 THE 30-DAY FETCH LOGIC (Memory)
    async def get_recent_history(self, client_id: str, days: int = 30) -> List[Dict[str, Any]]:
        db = await self.get_db()
        cutoff = datetime.now() - timedelta(days=days)
        
        # Fetch past sessions and insights for this client
        cursor = db.ai_sessions.find({
            "clientId": client_id,
            "createdAt": {"$gte": cutoff}
        }).sort("createdAt", -1).limit(10)
        
        history = await cursor.to_list(length=10)
        return history

